src/tarantool.client/ResponseReader.cs

```diff
@@ -29,14 +35,65 @@ internal class ResponseReader : IResponseReader
 
         private bool _disposed;
 
-        public ResponseReader(ILogicalConnection logicalConnection, ClientOptions clientOptions, INetworkStreamPhysicalConnection physicalConnection)
+        public ResponseReader(ClientOptions clientOptions, INetworkStreamPhysicalConnection physicalConnection)
         {
             _physicalConnection = physicalConnection;
-            _logicalConnection = logicalConnection;
             _clientOptions = clientOptions;
             _buffer = new byte[clientOptions.ConnectionOptions.ReadStreamBufferSize];
         }
 
+        public Task<MemoryStream> GetResponseTask(RequestId requestId)
+        {
+            try
+            {
+                _pendingRequestsLock.EnterWriteLock();
+
+                if (_disposed)
+                {
+                    throw new ObjectDisposedException(nameof(ResponseReader));
+                }
+
+                if (_pendingRequests.ContainsKey(requestId))
+                {
+                    throw ExceptionHelper.RequestWithSuchIdAlreadySent(requestId);
+                }
+
+                var tcs = new TaskCompletionSource<MemoryStream>();
+                _pendingRequests.Add(requestId, tcs);
+
+                return tcs.Task;
+            }
+            finally
+            {
+                _pendingRequestsLock.ExitWriteLock();
+            }
+        }
+
+        private TaskCompletionSource<MemoryStream> PopResponseCompletionSource(RequestId requestId, MemoryStream resultStream)
+        {
+            try
+            {
+                _pendingRequestsLock.EnterWriteLock();
+
+                if (_disposed)
+                {
+                    throw new ObjectDisposedException(nameof(ResponseReader));
+                }
+
+                TaskCompletionSource<MemoryStream> request;
+                if (_pendingRequests.TryGetValue(requestId, out request))
+                {
+                    _pendingRequests.Remove(requestId);
```

Comment:
Remove does not throw exceptions if key is missing. We can remove check for TryGetValue.